src/main/java/bgu/spl/net/impl/stomp/LibraryProtocol.java

```diff
@@ -13,59 +13,168 @@
+            if (info != null) addErrorBody(msg, f, info);
+            _connections.send(_connectionId,
+                    f);
+            return false;
+        }
+        return true;
+    }
+
+    private boolean checkForHeader(Frame msg, String header) {
+        return checkForHeader(msg, header, null);
+    }
+
+    private void handleSubscribe(Frame msg) {
+        if (!checkForHeader(msg, "destination")) return;
+        if (!checkForHeader(msg, "id")) return;
+        _connections.subscribe(msg.getHeader("destination"), _connectionId, Integer.parseInt(msg.getHeader("id")));
+        sendReceipt(msg);
+    }
+
+    private void handleUnsubscribe(Frame msg) {
+        if (!checkForHeader(msg, "id")) return;
+        try {
+            _connections.unsubscribe(msg.getHeader("destination"), _connectionId, Integer.parseInt(msg.getHeader("id")));
+        }
+        // We ignore a NullPointer exception because one will be thrown if the subscriptionId doesn't exist
+        catch (NullPointerException ignored) {
+        }
+        sendReceipt(msg);
+    }
+
+    /**
+     * Adds a receipt id needed according to {@code msg}'s headers
+     *
      * @param msg the received message
-     * @param newFrame the currently building frame
-     * @return the newFrame with the receipt, this is also in-place but this is useful
      */
-    private Frame addReceipt(Frame msg, Frame newFrame) {
+    private void sendReceipt(Frame msg) {
         String receiptId = msg.getHeader("receipt");
-        if(receiptId != null) {
-            newFrame.addHeader("receiptId", "message-"+receiptId);
+        if (receiptId != null) {
+            Frame newFrame = new Frame("RECEIPT");
+            newFrame.addHeader("receipt-id", "message-" + receiptId);
+            _connections.send(_connectionId, newFrame);
         }
-        return newFrame;
+    }
+
+    private Frame createErrorFrame(String message) {
+        Frame f = new Frame("ERROR");
+        f.addHeader("message", message);
+        return f;
+    }
+
+    /**
+     * Creates a message frame. Note that this doesn't add the subscription-id, which will be added by the
+     * Connections interface once the message gets sent
+     *
+     * @param msg The message received by the client which triggered the call to this function
+     * @return The message to be sent to the users in the channel (without the subscription-id)
+     */
+    private Frame createMessageFrame(Frame msg) {
+        Frame f = new Frame("MESSAGE");
+        f.addHeader("Message-id", _messagesSent.toString());
+        _messagesSent++;
+        f.addHeader("destination", msg.getHeader("destination"));
+        f.addBody(msg.getBody());
+        return f;
     }
 
     private void handleSend(Frame msg) {
-        // TODO: Complete
+        if (!checkForHeader(msg, "destination")) return;
+        _connections.send(msg.getHeader("destination"), createMessageFrame(msg));
+        sendReceipt(msg);
```

Comment:
no receipt needed for SEND in my opinion

Reply:
In the assignment specification they state that any frame may contain a receipt id and a receipt should be sent if it does

Reply:
I missed that you're sending only if there is need to.
approved (: